symengine/uint_base.h

```diff
@@ -0,0 +1,91 @@
+/**
+ *  \file polynomial_int.h
+ *  Class for Univariate Polynomial Base
+ **/
+#ifndef SYMENGINE_UINT_BASE_H
+#define SYMENGINE_UINT_BASE_H
+
+#include <symengine/basic.h>
+
+namespace SymEngine
+{
+
+template <typename Container, typename Poly>
+class UIntPolyBase : public Basic
+{
+protected:
+    RCP<const Symbol> var_;
+    Container poly_;
+
+public:
+    UIntPolyBase(const RCP<const Symbol> &var, Container &&container)
+        : var_{var}, poly_{container}
+    {
+    }
+
+    // unify these two constructor? another template would be required
+    // may solve some more problems, like `get_degree` virtualization
+    UIntPolyBase(const RCP<const Symbol> &var,
+                 const std::vector<integer_class> &v)
+        : var_{var}
+    {
+    }
+
+    UIntPolyBase(const RCP<const Symbol> &var, const std::vector<Expression> &v)
+        : var_{var}
+    {
+    }
+
+    // TODO think of something to make this purely virtual
+    //! \returns the degree of the polynomial
+    // virtual unsigned int get_degree() const = 0;
+
+    //! \returns `-1`,`0` or `1` after comparing
+    virtual int compare(const Basic &o) const = 0;
+
+    //! \returns `true` if two objects are equal
+    inline bool __eq__(const Basic &o) const
+    {
+        return eq(*var_, *(static_cast<const Poly &>(o).var_))
+               and poly_ == static_cast<const Poly &>(o).poly_;
+    }
+
+    inline RCP<const Symbol> get_var() const
+    {
+        return var_;
+    }
+
+    inline const Container &get_poly() const
+    {
+        return poly_;
+    }
+
+    friend RCP<const Poly> add_poly(const Poly &a, const Poly &b)
```

Comment:
If it's not a function inside the class, I will have to make two separate functions for both the classes. This will beat the purpose of the template, I think.

Even if I templatize say `add_poly<Poly, Container>` I'll have to derive `add_poly_int` and `add_poly_expr` to make it work.